exercises/robot-name/examples/success-standard/src/Robot.hs

```diff
@@ -1,22 +1,43 @@
-module Robot (Robot, mkRobot, robotName, resetName) where
+module Robot (Robot, mkRobot, resetName, robotName, initialState) where
 
-import Control.Concurrent (MVar, readMVar, swapMVar, newMVar)
-import Control.Monad (void)
-import System.Random (randomRIO)
+import           Control.Concurrent.MVar (MVar, newMVar, putMVar, readMVar,
+                                          takeMVar)
+import           Control.Monad.State     (StateT, get, modify)
+import           Control.Monad.Trans     (lift)
+import           Data.Set                (Set)
+import qualified Data.Set                as S
+import           System.Random           (randomRIO)
 
 newtype Robot = Robot { robotNameVar :: MVar String }
+type RunState = Set String
 
-mkRobot :: IO Robot
-mkRobot = fmap Robot $ generateName >>= newMVar
+initialState :: RunState
+initialState = S.empty
+
+randomName :: IO String
+randomName = mapM randomRIO [letter, letter, digit, digit, digit]
+  where
+    letter = ('A', 'Z')
+    digit = ('0', '9')
+
+randomUniqueName :: StateT RunState IO String
+randomUniqueName = do
+  name <- lift randomName
+  nameSpace <- get
+  if name `S.member` nameSpace
+    then randomUniqueName
+    else return name
+
+mkRobot :: StateT RunState IO Robot
+mkRobot = fmap Robot $ randomUniqueName >>= lift . newMVar
+
+resetName :: Robot -> StateT RunState IO ()
+resetName (Robot name) = do
+  oldName <- lift $ takeMVar name
+  newName <- randomUniqueName
+  modify (S.delete oldName . S.insert newName)
```

Comment:
The specs for the exercise seem to be misleading.  I thought that every name had to be unique, even if they were reset (i.e. never delete the old names).  This is obviously a different interpretation.  I don't know if there is a "correct" interpretation?